### PR TITLE
[SYCL] Add esimd device descriptor for 2d load/store/prefetch

### PR DESCRIPTION
**Pull request by againull**
Add esimd device descriptor to check if 2d block operations are supported by the device. 
UR counterpart: https://github.com/oneapi-src/unified-runtime/pull/2261

### CODE REVIEW

**Comment by Pennycook** [comment]
Am I right in thinking that this feature (2D block loads/stores) are not actually exposed to SYCL?

**Comment by rolandschulz** [comment]
It is indirectly available to users through the joint_matrix abstraction. But for that I don't think the user needs to be able to query it. Currently it isn't available directly in a supported way (cutlass uses IGC builtins). But we plan to support it at least through SPIR-V/OpenCL interop and if there demand also as SYCL extension.

**Comment by gmlueck** [comment]
It does seem weird to add a SYCL query for a feature that cannot be used via SYCL.  Can we instead add the query at the same level as the API?  For example, if applications will access 2D block load/store via OpenCL, can we add an OpenCL query?

**Comment by againull** [comment]
According to the info from the user they use either esimd  or joint_matrix.
So, I believe if someone uses joint_matrix operations then they can check corresponding matrix aspects.

In case of ESIMD it looks like load_2d/store_2d operations are mapped to this low level feature https://github.com/intel/compute-runtime/blob/24.39.31294.12/level_zero/doc/experimental_extensions/2D_BLOCK_TRANSPOSE.md.
@sarnex  Could you please help to confirm that or maybe recommend someone who can help.

Currently in the ESIMD docs there is a statement that load_2d/store_2d can be used only on PVC (see https://github.com/intel/llvm/pull/15905/files#diff-fef71b0090d0ea44c96240cd9d6497881b617b91e234635a19313eb69864b7f3L555): 
I believe we can create a new esimd  aspect (which checks if device has 2d block load/store) for this load_2d/store_2d operations.
So that users don't have to check particular devices but can just check an aspect.

Does it make sense to introduce such aspect? I've updated the PR with proposed changes.

**Comment by sarnex** [comment]
I can't say for sure if the architecture restriction is correct (idk if 2d ops are PVC only, or actually PVC or later or what).
But I have no flags to adding an aspect, I guess it would just come down to a arch/device ID check in UR somewhere.

**Comment by againull** [comment]
Changed implementation to device descriptor in esimd namespace based on offline discussion.

File: sycl/test/abi/sycl_symbols_windows.dump [hunk 1]
```diff
@@ -156,6 +157,7 @@
 ??$get_info_impl@Ugpu_slices@device@info@intel@ext@_V1@sycl@@@device@_V1@sycl@@AEBAIXZ
 ??$get_info_impl@Ugpu_subslices_per_slice@device@info@intel@ext@_V1@sycl@@@device@_V1@sycl@@AEBAIXZ
 ??$get_info_impl@Uhalf_fp_config@device@info@_V1@sycl@@@device@_V1@sycl@@AEBA?AV?$vector@W4fp_config@info@_V1@sycl@@V?$allocator@W4fp_config@info@_V1@sycl@@@std@@@std@@XZ
+??$get_info_impl@Uhas_2d_block_io_support@device@info@esimd@intel@ext@_V1@sycl@@@device@_V1@sycl@@AEBA_NXZ
```
Comment (sarnex) [inline sycl/test/abi/sycl_symbols_windows.dump:160]: this shouldnt be abi breaking right?

File: sycl/doc/extensions/supported/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md [hunk 2]
```diff
@@ -1018,6 +1019,11 @@ The new aspect has the following behavior when queried via `device::has()`:
 |--------|-------------|
 |`aspect::ext_intel_esimd` | Indicates that the device supports the `sycl_ext_intel_esimd` extension as defined in this document. |
 
+## Device Information Descriptors
+| Device Descriptors | Return Type | Description |
+| ------------------ | ----------- | ----------- |
+| `ext::intel::esimd::info::device::has_2d_block_io_support` | bool | Returns the boolean indicating whether 2D load/store/prefetch instructions are supported by the device. |
```
Comment (sarnex) [inline sycl/doc/extensions/supported/sycl_ext_intel_esimd/sycl_ext_intel_esimd.md:1025]:
```suggestion
| `ext::intel::esimd::info::device::has_2d_block_io_support` | bool | Returns a boolean indicating whether 2D load/store/prefetch instructions are supported by the device. |
```

**Review by sarnex — APPROVED** [review]
lgtm!

File: sycl/test/abi/sycl_symbols_windows.dump (hunk 1 above)
Comment (againull) [reply sycl/test/abi/sycl_symbols_windows.dump:160]: Yes, adding a new symbol is not a breaking change.

**Comment by againull** [comment]
dyn_cgf_different_arg_nums.cpp failure on CUDA is unrelated and will be fixed here: https://github.com/intel/llvm/pull/16156

**Review by cperkinsintel — APPROVED** [review]

**Comment by againull** [comment]
@intel/unified-runtime-reviewers  Could you please approve.

**Review by callumfare — APPROVED** [review]

**Comment by AlexeySachkov** [comment]
@againull, FYI, this PR breaks some tests on PVC:
- ESIMD/dpas/dpas_tf32.cpp
- ESIMD/acc_gather_scatter_rgba_stateless_64.cpp
- ESIMD/lsc/lsc_block_load_store_stateless_flag_64.cpp
- ESIMD/lsc/lsc_block_load_store_stateless_64.cpp
- ESIMD/accessor_stateless_ctor_64.cpp
- ESIMD/accessor_load_store_stateless_64.cpp
- ESIMD/lsc/lsc_gather_scatter_stateless_64.cpp
- ESIMD/lsc/lsc_load_store_2d_compare.cpp
- ESIMD/accessor_gather_scatter_stateless_64.cpp
- ESIMD/accessor_stateless_64.cpp
- ESIMD/slm_init_no_inline.cpp

**Comment by againull** [comment]
> @againull, FYI, this PR breaks some tests on PVC:
> 
> * ESIMD/dpas/dpas_tf32.cpp
> * ESIMD/acc_gather_scatter_rgba_stateless_64.cpp
> * ESIMD/lsc/lsc_block_load_store_stateless_flag_64.cpp
> * ESIMD/lsc/lsc_block_load_store_stateless_64.cpp
> * ESIMD/accessor_stateless_ctor_64.cpp
> * ESIMD/accessor_load_store_stateless_64.cpp
> * ESIMD/lsc/lsc_gather_scatter_stateless_64.cpp
> * ESIMD/lsc/lsc_load_store_2d_compare.cpp
> * ESIMD/accessor_gather_scatter_stateless_64.cpp
> * ESIMD/accessor_stateless_64.cpp
> * ESIMD/slm_init_no_inline.cpp

Are there any logs available for these failures?

**Comment by againull** [comment]
> @againull, FYI, this PR breaks some tests on PVC:
> 
> * ESIMD/dpas/dpas_tf32.cpp
> * ESIMD/acc_gather_scatter_rgba_stateless_64.cpp
> * ESIMD/lsc/lsc_block_load_store_stateless_flag_64.cpp
> * ESIMD/lsc/lsc_block_load_store_stateless_64.cpp
> * ESIMD/accessor_stateless_ctor_64.cpp
> * ESIMD/accessor_load_store_stateless_64.cpp
> * ESIMD/lsc/lsc_gather_scatter_stateless_64.cpp
> * ESIMD/lsc/lsc_load_store_2d_compare.cpp
> * ESIMD/accessor_gather_scatter_stateless_64.cpp
> * ESIMD/accessor_stateless_64.cpp
> * ESIMD/slm_init_no_inline.cpp

Thanks, those have requires: pvc, so I didn't get failures in pre/post-commit.  I know the rootcause, will fix.

**Pull request merged**